core/src/main/java/org/elasticsearch/indices/recovery/RecoverySource.java

```diff
@@ -200,32 +177,50 @@ synchronized void cancel(IndexShard shard, String reason) {
+             * Adds recovery source handler if recoveries are not blocked from starting (see also {@link #blockNewRecoveries}.
              * Throws {@link DelayRecoveryException} if new recoveries are blocked from starting.
              */
-            synchronized RecoverySourceHandler addNewRecovery(Supplier<RecoverySourceHandler> recoverySourceHandlerSupplier) {
+            synchronized RecoverySourceHandler addNewRecovery(StartRecoveryRequest request, IndexShard shard) {
                 if (onNewRecoveryException != null) {
                     throw onNewRecoveryException;
                 }
-                RecoverySourceHandler handler = recoverySourceHandlerSupplier.get();
+                RecoverySourceHandler handler = createRecoverySourceHandler(request, shard);
                 recoveryHandlers.add(handler);
                 return handler;
             }
 
+            private RecoverySourceHandler createRecoverySourceHandler(StartRecoveryRequest request, IndexShard shard) {
+                RecoverySourceHandler handler;
+                final RemoteRecoveryTargetHandler recoveryTarget =
+                    new RemoteRecoveryTargetHandler(request.recoveryId(), request.shardId(), transportService, request.targetNode(),
+                        recoverySettings, throttleTime -> shard.recoveryStats().addThrottleTime(throttleTime));
+                Supplier<Long> currentClusterStateVersionSupplier = () -> clusterService.state().getVersion();
+                if (shard.indexSettings().isOnSharedFilesystem()) {
+                    handler = new SharedFSRecoverySourceHandler(shard, recoveryTarget, request, currentClusterStateVersionSupplier,
+                        this::blockNewRecoveries, logger);
+                } else {
+                    handler = new RecoverySourceHandler(shard, recoveryTarget, request, currentClusterStateVersionSupplier,
+                        this::blockNewRecoveries, recoverySettings.getChunkSize().bytesAsInt(), logger);
+                }
+                return handler;
+            }
+
             /**
-             * Makes new recoveries throw the provided exception. If set to null, re-enables starting new recoveries.
+             * Makes new recoveries throw a {@link DelayRecoveryException} with the provided message.
              *
-             * Throws {@link IllegalStateException} if new recoveries are already being prevented or allowed.
+             * Throws {@link IllegalStateException} if new recoveries are already being blocked.
              */
-            synchronized void throwExceptionOnNewRecovery(@Nullable DelayRecoveryException onNewRecoveryException)
-                throws IllegalStateException {
-                if (onNewRecoveryException != null && this.onNewRecoveryException != null) {
+            synchronized Releasable blockNewRecoveries(String exceptionMessage) throws IllegalStateException {
```

Comment:
nit: block suggest "block the thread" like on have it wait on a lock. don't you like "delay"?